### PR TITLE
SearchKit - Editable refresh/performance/permission improvements

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Multiple improvements to the inline-edit feature.

- Editable fields can skip permission checks when the searchDisplay is configured for `acl_bypass`
- Option lists are pre-loaded and don't need to be fetched prior to editing a field
- Data is saved & the entire row refreshed server-side in a single ajax request
- Editable metadata is loaded once per column instead of once per row, improving performance
- Save & refresh are non-blocking and another field can be edited simultaneously
- Improve test coverage

Comments
----------------------------------------
TODO: Keyboard support for tab/arrow key navigation to and from editable fields (see https://lab.civicrm.org/dev/core/-/issues/4632 )

### CODE REVIEW

**Comment by allinappliadmin** [comment]
I've tested the patch on 5.78.4 and the output of a SK dashlets changed. No data are shown anymore when it should. Before the patch I was getting results from SearchKit

image

**Comment by colemanw** [comment]
@allinappliadmin that sounds like a server-side error. To help determine the cause, could you please open your browser console and check for error messages?

**Comment by allinappliadmin** [comment]
actually I was also debugging with Nicol so I had to revert the patch to check the style on SK. Do you need me to apply again the patch and see how it runs?

**Comment by colemanw** [comment]
@allinappliadmin yes please see if you can find the error

**Comment by allinappliadmin** [comment]
I've applied the patch but it has thrown this error:
```
Détails de l’erreur
======================
Une erreur de type E_COMPILE_ERROR a été causée dans la ligne 43 du fichier /var/www/vhosts/.../httpdocs/wp-content/plugins/civicrm/civicrm/ext/search_kit/Civi/Api4/Action/SearchDisplay/Run.php. Message d’erreur : Could not check compatibility between Civi\Api4\Action\SearchDisplay\Run::processResult(Civi\Api4\Action\SearchDisplay\SearchDisplayRunResult $result) and Civi\Api4\Action\SearchDisplay\AbstractRunAction::processResult(Civi\Api4\Result\SearchDisplayRunResult $result), because class Civi\Api4\Action\SearchDisplay\SearchDisplayRunResult is not available
```

**Comment by colemanw** [comment]
@allinappliadmin sounds like the patch was incorrectly applied

**Comment by colemanw** [comment]
@allinappliadmin if you can't apply the patch you can always test this on the [online demo site](http://site-list.test-1.civicrm.org:8001/?filter=core-31382-*)

**Comment by allinappliadmin** [comment]
I've applied several times the patch like any other and it's just breaking our test instance and the test site doesn't work either :-/

**Comment by colemanw** [comment]
@allinappliadmin yes I see building the demo site has the same problem:
```
Applying: SearchKit - Editable refresh/performance/permission improvements   
Patch failed at 0001 SearchKit - Editable refresh/performance/permission improvements
```

Git sometimes has trouble applying patches when nearby parts of a file have also been recently modified. When you apply the patch, look for the error message and then manually edit that file.

I will rebase this PR and update the demo site in an hour or two after resolving another issue and that will give you an online demo you can test here, and possibly a cleaner patch to use on your local site.

**Comment by colemanw** [comment]
@allinappliadmin I've rebased this PR, and the online demo is now working. The patch should now apply cleanly to `master`.

**Comment by allinappliadmin** [comment]
the patch doesn't get correctly applied on my side :-/

**Comment by allinappliadmin** [comment]
I can access the test site but FormBuilder is not installed

image

**Comment by allinappliadmin** [comment]
what would help is to have a set of out-of-the-box SK/FB also on the demo/test sites

**Comment by allinappliadmin** [comment]
@UshaMakoa were you able to test this patch?

**Comment by colemanw** [comment]
@allinappliadmin I agree it would be nice for FormBuilder to be pre-installed on demo sites, but you do have admin access and can install it with a few clicks.
Here, I have done it for you. Try the demo site again.

**Comment by colemanw** [comment]
@allinappliadmin FYI the admin login is here:

---

Screenshot 2024-11-06 at 9 14 02 AM

**Comment by allinappliadmin** [comment]
from what I've just tried here : http://core-31382-2pgo7.test-1.civicrm.org:8003/civicrm/test clicking on 'tab' button on my keyboard switches from one tick box to an other i.e vertically but not horizontally. I also don't know what was actually expected but at least the keyboard controls something.

With inline edit on a field clicking on tab selects the value instead of needing the select the green check sign, also cool

**Comment by colemanw** [comment]
@allinappliadmin to clarify, this PR has nothing to do with tabs. Please do not test tabs as part of your review of this PR.
The scope of this PR is detailed in the **Overview** section above.

**Comment by allinappliadmin** [comment]
oups sorry

**Comment by allinappliadmin** [comment]
happy to test something new :-)

**Comment by allinappliadmin** [comment]
not sure that my data set was the proper one to do these tests but I would say that it works

**Comment by allinappliadmin** [comment]
not really helpful...

**Comment by angusfretwell** [comment]
Really keen on this for the ACL bypass on editable fields. Let me know if I can help get this over the line with testing etc. 😄

**Comment by colemanw** [comment]
@angusfretwell absolutely! This PR is ready to be merged, just waiting for someone to do a review. Anyone can do it. See the "Quick links for reviewers" just below the description.

**Comment by angusfretwell** [comment]
test this please

**Comment by angusfretwell** [comment]
@colemanw Tried to get the online demo going, I think this needs a rebase so patch can be applied.

**Comment by colemanw** [comment]
Thanks @angusfretwell. Rebased.

**Comment by angusfretwell** [comment]
@colemanw skipping permissions for inline editing doesn't seem to be working. In my testing I've observed ACLs still applying, e.g. for a user that has edit permissions for a limited set of contacts, they can only edit those contacts but not others.

I've also observed on front-end forms that search displays that have inline edit fields won't render anything for a user without the "CiviCRM: access CiviCRM backend and API" Drupal permission. I'm pretty sure the current behaviour is that it'll render the results. 

Just checking the only thing I need to toggle is the ACL bypass here?

image

File: ext/search_kit/ang/crmSearchDisplay/crmSearchDisplayEditable.component.js [hunk 1]
```diff
@@ -7,43 +7,53 @@
...
+          serialize: editableInfo.serialize,
+          nullable: editableInfo.nullable && ctrl.row.data[editableInfo.id_path],
         };
 
+        if (this.field.options === true) {
+          loadOptions();
+        }
+
         $(document).on('keydown.crmSearchDisplayEditable', (e) => {
+          console.log('keydown', e.key);
```
Comment (angusfretwell) [inline ext/search_kit/ang/crmSearchDisplay/crmSearchDisplayEditable.component.js:43]: 🧹

**Comment by angusfretwell** [comment]
>Option lists are pre-loaded and don't need to be fetched prior to editing a field
Works as expected

>Data is saved & the entire row refreshed server-side in a single ajax request
Works as expected

>Save & refresh are non-blocking and another field can be edited simultaneously
Works as expected

File: ext/search_kit/ang/crmSearchDisplay/crmSearchDisplayEditable.component.js (hunk 1 above)
Comment (colemanw) [reply ext/search_kit/ang/crmSearchDisplay/crmSearchDisplayEditable.component.js]: Thanks. Cleaned.

**Comment by colemanw** [comment]
Thanks @angusfretwell I think this is `merge-ready` based on your review.

> Is this an optimisation loading from the database? Not sure how to verify this one.

The json response coming back from loading search results (`Api4.searchDisplay.run`) is now a bit smaller because it removes the redundant `editable` info from every row. Verifying that it still works before/after is the important part, which you've done.

**Comment by angusfretwell** [comment]
@colemanw sorry for the slow response! Per [this comment](https://github.com/civicrm/civicrm-core/pull/31382#issuecomment-2484806130), I wasn't able to get the ACL bypass for inline edit working. Could you maybe provide some setup steps so I can re-test it?

**Comment by colemanw** [comment]
@angusfretwell I've triggered the demo site to rebuild, so you can test online now.

**Pull request merged**